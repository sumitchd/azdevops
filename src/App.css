.pl-5 {
  padding-left: 5px;
}
.App-header {
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  font-size: calc(2px + 2vmin);
  color: white;
}

.App-link {
  color: #61dafb;
}

.input {
  height: 30px;
  position: relative;
  border: 2px solid #cdcdcd;
  border-color: rgba(0, 0, 0, 0.14);
  font-size: 18px;
  width: 400px;
}

.btn {
  width: 150px;
  padding: 10px;
  border: 4px double;
  margin: 10px;
  margin-left: 0px;
}
