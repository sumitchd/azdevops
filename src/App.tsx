import React, { useState } from "react";
import { IUser } from "./models/core-services.model";
import { AppContext } from "./services/app-context";
import "./App.css";
import { AzureDevopsService } from "./services/azure-devops.service";
import {
  IItem,
  ITreeView,
  IWorkItemIds,
  IWorkItemsWithRelations,
} from "./models/app.model";
import { appConstants, uiMessages } from "./constants/ui.constants";

function App() {
  const { authClient, httpClient } = React.useContext(AppContext);
  const azDevopsService = new AzureDevopsService(httpClient);
  const [user, setUser] = React.useState<IUser | null>(null);
  const [reloadData, setReloadData] = React.useState(true);
  const [treeData, setTreeData] = React.useState(Array<ITreeView>());
  const [currentEditItem, setCurrentEditItem] = useState({} as IItem);
  const [pageError, setPageError] = useState("");
  const orgName = appConstants.orgName;
  const projectName = appConstants.projectName;
  React.useEffect(() => {
    (async () => {
      if (!user) {
        try {
          const user = await authClient.getUser();
          user ? setUser(user) : authClient.login().catch();
        } catch {
          authClient.login();
        }
      } else {
        if (reloadData) {
          try {
            const response: IWorkItemsWithRelations = await azDevopsService.GetWorkItemIDsAsync(
              orgName,
              projectName
            );
            const commaSeperatedIds = response.workItemRelations
              .map((x) => x.target?.id)
              .join();

            const itemDetailResponse: IWorkItemIds = await azDevopsService.GetWorkItemDetailsByIdAsync(
              orgName,
              projectName,
              commaSeperatedIds
            );
            const treeData = azDevopsService.createHierarchy(
              response.workItemRelations,
              itemDetailResponse.value
            );
            setTreeData(treeData);
            setReloadData(false);
          } catch {
            setPageError(uiMessages.pageError);
          }
        }
      }
    })();
  }, [user, reloadData]);

  const onEditClick = (id: number, currentTitle: string) => {
    setCurrentEditItem({ id, title: currentTitle });
    setTimeout(() => {
      const inputbox = document.getElementById(
        "editTextbox"
      ) as HTMLInputElement;
      if (inputbox) {
        inputbox.value = currentTitle;
        inputbox?.focus();
      }
    }, 0);
  };

  const onUpdateClick = async () => {
    const inputEle = document.getElementById("editTextbox") as HTMLInputElement;
    if (inputEle?.value) {
      const body = [
        {
          op: "replace",
          path: "/fields/System.Title",
          value: inputEle.value,
        },
      ];
      const response = await azDevopsService.UpdateWorkItemFieldAsync(
        orgName,
        projectName,
        body,
        currentEditItem.id
      );
      if (response) {
        setCurrentEditItem({} as IItem);
        inputEle.value = "";
        setReloadData(true);
        alert(uiMessages.updateSuccess);
      }
    }
  };

  const printWITree = (data: ITreeView[]) => {
    return data.map((item) => {
      return (
        <ul key={`ul_${item.id}`}>
          <li key={`li_${item.id}`}>
            <span>
              {`${uiMessages.id}: ${item.id} ${uiMessages.title}: ${item.title} `}
              <a
                href="#"
                className="App-link"
                data-testid={"edit" + item.id}
                id={item.id.toString()}
                onClick={(e) => {
                  onEditClick(+e.currentTarget.id, item.title);
                }}
              >
                {uiMessages.edit}
              </a>
            </span>
          </li>
          {item.children?.length ? printWITree(item.children) : null}
        </ul>
      );
    });
  };

  return (
    <div className="pl-5 App-header">
      <p>
        {`${uiMessages.hello}, ${user?.name}`}!
        <a
          href="#"
          className="pl-5 App-link"
          onClick={async () => {
            await authClient.logOut();
          }}
        >
          {uiMessages.logout}
        </a>
      </p>
      {pageError ? pageError : printWITree(treeData)}
      <div>
        {currentEditItem.id ? (
          <span>
            <input
              type="text"
              id="editTextbox"
              placeholder={uiMessages.editItemText}
              className="input"
              data-testid="editTextbox"
            ></input>
            <br />
            <button
              className="btn"
              data-testid="btnUpdate"
              onClick={onUpdateClick}
            >
              {uiMessages.update}
            </button>
          </span>
        ) : null}
      </div>
    </div>
  );
}

export default App;
