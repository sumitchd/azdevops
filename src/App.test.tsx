import React from "react";
import { render, screen } from "@testing-library/react";
import App from "./App";

test("renders hello and updateTextbox", () => {
  render(<App />);
  const helloText = screen.getByText(/Hello/i);
  expect(helloText).toBeInTheDocument();
});

test("renders with WI data", () => {
  React.useState = jest
    .fn()
    .mockReturnValueOnce([{ name: "Sumit" }, {}])
    .mockReturnValueOnce([false, {}])
    .mockReturnValueOnce([
      [
        {
          id: 1,
          title: "new item",
          children: [
            { id: 2, title: "new child item" },
            { id: 3, title: "new child item 2" },
            {
              id: 4,
              title: "new child item 4",
              children: [
                { id: 5, title: "new child item 5" },
                { id: 6, title: "new child item 6" },
              ],
            },
          ],
        },
      ],
      {},
    ])
    .mockReturnValueOnce([{ id: 3, title: "new child item 3" }]);
  const appTree = render(<App />);
  expect(appTree.container).toMatchSnapshot();
});
