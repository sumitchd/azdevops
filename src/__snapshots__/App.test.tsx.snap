// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with WI data 1`] = `
<div>
  <div
    class="pl-5 App-header"
  >
    <p>
      Hello Sumit
      !
      <a
        class="pl-5 App-link"
        href="#"
      >
        Logout
      </a>
    </p>
    <ul>
      <li>
        <span>
          ID: 1 Title: new item 
          <a
            class="App-link"
            data-testid="edit1"
            href="#"
            id="1"
          >
            Edit
          </a>
        </span>
      </li>
      <ul>
        <li>
          <span>
            ID: 2 Title: new child item 
            <a
              class="App-link"
              data-testid="edit2"
              href="#"
              id="2"
            >
              Edit
            </a>
          </span>
        </li>
      </ul>
      <ul>
        <li>
          <span>
            ID: 3 Title: new child item 2 
            <a
              class="App-link"
              data-testid="edit3"
              href="#"
              id="3"
            >
              Edit
            </a>
          </span>
        </li>
      </ul>
      <ul>
        <li>
          <span>
            ID: 4 Title: new child item 4 
            <a
              class="App-link"
              data-testid="edit4"
              href="#"
              id="4"
            >
              Edit
            </a>
          </span>
        </li>
        <ul>
          <li>
            <span>
              ID: 5 Title: new child item 5 
              <a
                class="App-link"
                data-testid="edit5"
                href="#"
                id="5"
              >
                Edit
              </a>
            </span>
          </li>
        </ul>
        <ul>
          <li>
            <span>
              ID: 6 Title: new child item 6 
              <a
                class="App-link"
                data-testid="edit6"
                href="#"
                id="6"
              >
                Edit
              </a>
            </span>
          </li>
        </ul>
      </ul>
    </ul>
    <div />
  </div>
</div>
`;
